javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -0,0 +1,164 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.io.Serializable;
+import java.util.function.Function;
+
+import static javaslang.collection.Arrays.*;
+
+final class BitMappedTrie<T> implements Serializable {
+    static int BRANCHING_BASE = 5;
+    static int branchingFactor()                   { return 1 << BRANCHING_BASE; }
+    static int firstDigit(int num, int depthShift) { return num >> depthShift; }
+    static int digit(int num, int depthShift)      { return lastDigit(firstDigit(num, depthShift)); }
+    static int lastDigit(int num)                  { return num & (-1 >>> -BRANCHING_BASE); }
+
+    private static final long serialVersionUID = 1L;
+
+    private static final BitMappedTrie<?> EMPTY = new BitMappedTrie<>(emptyArray(), 0, 0, 0);
+
+    @SuppressWarnings("unchecked")
+    static <T> BitMappedTrie<T> emptyTree() { return (BitMappedTrie<T>) EMPTY; }
+
+    private final Object[] array;
+    private final int offset, length;
+    private final int depthShift;
+
+    BitMappedTrie(Object[] array, int offset, int length, int depthShift) {
+        this.array = array;
+        this.offset = offset;
+        this.length = length;
+        this.depthShift = depthShift;
+
+        assert length() <= treeSize(branchingFactor(), depthShift);
+    }
+
+    private static int treeSize(int branchCount, int depthShift) {
+        final int fullBranchSize = 1 << depthShift;
+        return branchCount * fullBranchSize;
+    }
+
+    BitMappedTrie<T> prependLeaf(T leading) {
```

Comment:
Why is this method called `prependLeaf` as opposed to just `prepend` ? Does it make sense to have, or can you imagine, an operation in which we prepend anyplace else but to the first leaf?